DirectProgramming/DPC++FPGA/Tutorials/Features/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,234 @@
+
+  q.submit([&](handler &h) {
+    accessor a_acc(a_buf, h, read_only);
+    accessor b_acc(b_buf, h, read_only);
+    accessor c_acc(c_buf, h, write_only, no_init);
+    h.single_task<EfficientShiftOp>([=]() [[intel::kernel_args_restrict]] {
+      c_acc[0] = a_acc[0] << b_acc[0];
+    });
+  });
+}
+
+my_int14 TestBitOps(queue &q, const my_int14 &a) {
+  my_int14 res;
+  buffer<my_int14, 1> a_buf(&a, 1);
+  buffer<my_int14, 1> res_buf(&res, 1);
+
+  q.submit([&](handler &h) {
+    accessor a_acc(a_buf, h, read_only);
+    accessor res_acc(res_buf, h, write_only, no_init);
+    h.single_task<BitOps>([=]() [[intel::kernel_args_restrict]] {
+      my_int14 temp = a_acc[0].slc<14>(0);
+
+      res_acc[0] = 0;  // Must be initialized before bit operations, otherwise,
+                       // it will be undefined behavior
+
+      // 0b0 -> 0b1111101000 (decimal:1000)
+      res_acc[0].set_slc(0, temp);
+
+      // 0b1111101000 -> 0b1111101111
+      res_acc[0][2] = 1;
+      res_acc[0][1] = 1;
+      res_acc[0][0] = 1;
+    });
+  });
+  return res;
+}
+
+int main() {
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector device_selector;
+#else
+  ext::intel::fpga_selector device_selector;
+#endif
+
+  bool passed = true;
+
+  try {
+    queue q(device_selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    const int t1 = 1000, t2 = 2;
+
+    // Kernel `BasicOpsInt` contains native `int` type addition, multiplication,
+    // and division operations, while kernel `BasicOpsAcInt` contains `ac_int`
+    // type addition, multiplication, and division operations. By comparing
+    // these two kernels, you will find reduced width `ac_int` generates more
+    // efficient hardware than native `int`.
+    {
+      my_int14 a = t1, b = t2;
+      my_int15 c;
+      my_int28 d;
+      my_int15 e;
+      TestBasicOpsAcInt(q, a, b, c, d, e);
+
+      int _c, _d, _e;
+      TestBasicOpsInt(q, a, b, _c, _d, _e);
+
+      if (c != _c || d != _d || e != _e) {
+        std::cout << "Result mismatch!\n"
+                  << "Kernel BasicOpsInt:   addition = " << _c
+                  << ", multiplication = " << _d << ", division = " << _e
+                  << "\n"
+                  << "Kernel BasicOpsAcInt: addition = " << c
+                  << ", multiplication = " << d << ", division = " << e
+                  << "\n\n";
+        passed = false;
+      }
+    }
+
+    // Kernel `ShiftOp` contains an `ac_int` left shifter and the data type of
+    // the shift amount is a large width signed `ac_int`. On contrast, kernel
+    // `EfficientShiftOp` also contains an `ac_int` left shifter but the data
+    // type of the shift amount is a reduced width unsigned `ac_int`. By
+    // comparing these two kernels, you will find shift operations of `ac_int`
+    // can generate more efficient hardware if the amount to shift by is stored
+    // in a minimally sized unsigned `ac_int`.
+    {
```

Comment:
I don't think you need any of these scopes. The buffers are declared inside the functions you call, so they will go out of scope inside the function and copy the data back.

Reply:
This applies to your other scopes as well.

Reply:
I'm not worrying about the buffers.
Without these block scopes, all `ac_int` local variables passed to different kernels must have their unique names - that's a lot to create.
With these scopes, each kernel can have its own `input_a`, `input_b`, etc.

Reply:
yep yep! Good call :)